Add proposal spec for dotnet-stack tool (#1696)
dotnet/diagnostics: documentation/design-docs/dotnet-tools.md

    ps              Lists dotnet processes that can be attached to.
    list-profiles   Lists pre-built tracing profiles with a description of what providers and filters are in each profile.
    convert         Converts traces to alternate formats for use with alternate trace analysis tools

>     stack           Print the stack trace for every thread in teh target process

Reviewer: ```suggestion
    stack           Print the managed stack trace of every thread in the target process
```

      Writing:       ./trace.speedscope.json
      Conversion complete

### dotnet-stack

SYNOPSIS

    dotnet-stack [options] [command] [<args>]

OPTIONS

    --version
        Display the version of the dotnet-trace utility.

    -h, --help
        Show command line help

COMMANDS

>     report         Displays stack traces for the target process

Reviewer 2: nit: We've used the verb `collect` in our other tools. Perhaps switch for consistency's sake?

Reviewer 3: My sense of past usage has been this:

- collect - capture data data write it to a file for offline analysis
- monitor - capture data and stream it to the console interactively
- report - capture data and display it on the console non-interactively

We only used 'report' in one other place (dotnet gcdump) so the trend isn't as well established. It seems like it fits to me, but I might also be splitting hairs that casual users aren't going to think about?

Author: Noah's logic is what led me to use `report`.  At some point, I'm hoping we might have a `report` verb on `dotnet-{trace|counters}` as well, that will allow you to produce views of data _a la_ `perf`'s `perf script ...` syntax.

      Writing:       ./trace.speedscope.json
      Conversion complete

### dotnet-stack

SYNOPSIS

    dotnet-stack [options] [command] [<args>]

OPTIONS

    --version
        Display the version of the dotnet-trace utility.

    -h, --help
        Show command line help

COMMANDS

    report         Displays stack traces for the target process

REPORT

    dotnet-stack report -p|--process-id <pid>
                        [-t|--report-type <report-type>]
                        [-h|--help]

    Prints the managed stack from every thread in the target process

    -t, ---report-type <report-type>

>         The way to represent the stack data.  (default: standard)

Author: I added this only since it gives us an easy place to add additional views of stack data, e.g., an Async stack view.

Reviewer 3: Until we support more than one choice I'd suggest not creating the option.

    -h, --help
        Show command line help

    Examples:

>       > dotnet-trace stack -p 1234

Reviewer 3: this part is still on the old design : )

>       Stack for Thread (0x1):

Reviewer 3: Are these IDs the OS id? The small numbers suggest that it isn't but I think it should be.

Author: I just put random numbers rather than convert the original numbers to their hex values.  I intend for these to be the OS thread Ids.

          System.Private.CoreLib!System.Threading.ThreadHelper.ThreadStart()

      Stack for Thread (0x4):
          [Native Frames]
          System.Private.CoreLib!System.Threading.Thread.Sleep(System.TimeSpan)
          Microsoft.AspNetCore.Server.Kestrel.Core!Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.Heartbeat.TimerLoop()
          Microsoft.AspNetCore.Server.Kestrel.Core!Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure.Heartbeat.ctor(System.Object)
          System.Private.CoreLib!System.Threading.ThreadHelper.ThreadStart_Context(System.Object)
          System.Private.CoreLib!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)
          System.Private.CoreLib!System.Threading.ThreadHelper.ThreadStart(System.Object)

      Stack for Thread (0x5):
          [Native Frames]
          System.Net.Sockets!System.Net.Sockets.SocketAsyncEngine.EventLoop()
          System.Net.Sockets!System.Net.Sockets.SocketAsyncEngine.ctor( System.Object)

>           System.Private.CoreLib!System.Threading.ThreadHelper.ThreadStart(System.Object)

Author: I attempted to simplify the printed stacks and make them more C#-like.  I don't think I got some of the conversions correct, though.  That being said, I don't know how much work it will be to get this style output from TraceEvent.  Would people be okay with the default output from the initial commit?

Reviewer 3: No qualms from me : )

Author: I'll leave the proposal with the preferred output, and if the implementation differs, we can adjust it back to the default output.

      Writing:       ./trace.speedscope.json
      Conversion complete

> CONVERT

Reviewer: ```suggestion
STACK
```

    dotnet-trace stack -p|--process-id <pid>
                       [-h|--help]

    Prints the stack for every thread in the target process

    -h, --help
        Show command line help

    Examples:
      > dotnet-trace stack -p 1234

>       Stack for Thread (2207382):

Reviewer 3: Nit: I'd suggest hex for this thread id

>           UNMANAGED_CODE_TIME

Reviewer: This looks weird in the output. Should we replace this with [Native Frames]? Or somethings alike?

Author: This is output from TraceEvent's symbolicating logic.  It should be fairly simple to just check for this text and replace it before printing.

>           System.Private.CoreLib!System.Threading.ManualResetEventSlim.Wait(int32,value class System.Threading.CancellationToken)

Reviewer 3: Devs would probably appreciate more C#-like syntax formatting. Are we formatting this string in the tool or it is being produced by the runtime? I wouldn't want to do anything that takes a lot of dev effort but if the tool formats the string its probably not too hard to pretty some of this stuff up:
- C# keywords for primitive types
- eliminate class/value class terminology
- '.' instead of '+' for inner class
- List\<Foo> instead of List`1[Foo]

Async stacks probably also don't look great but improving them is likely a more substantial task that we wouldn't need to take on right away.

Author: > Are we formatting this string in the tool or it is being produced by the runtime?

This is the default output from the TraceEvent library I used for symbol resolution.  I haven't investigated symbolicating enough to know how expensive it would be to get a more C# style representation.

https://github.com/josalem/DotStack/blob/2ddc831df6f7aaa9111f8eb6246bb80ace896610/DotStack.cs#L103-L113
